converter/output/rita/db.go

```diff
@@ -0,0 +1,140 @@
+	"github.com/globalsign/mgo"
+	"github.com/pkg/errors"
+	"time"
+)
+
+//DB represents a RITA database and provides routines
+//for constructing a new RITA database and inserting
+//new data into that database
+type DB struct {
+	manager         DBManager
+	outputDB        *mgo.Database
+	connColl        *buffered.AutoFlushCollection
+	connCounter     freqconn.ConnCounter
+	strobesNotifier freqconn.StrobesNotifier
+}
+
+func newDB(dbManager DBManager, outputDB *mgo.Database,
+	strobeThreshold int, bufferSize int64, flushDeadline time.Duration,
+	asyncErrorChan chan<- error, onFatalError func()) (DB, error) {
+
+	strobesSess := outputDB.Session.Copy()
+	connSess := outputDB.Session.Copy()
+
+	connColl := buffered.NewAutoFlushCollection(
+		outputDB.C(constants.ConnCollection).With(connSess),
+		bufferSize, flushDeadline,
+	)
+
+	//The strobes notifier needs access to the connColl so it can flush
+	//the connColl buffer before it removes entries from the conn collection
+	strobesNotifier := freqconn.NewStrobesNotifier(outputDB.With(strobesSess), connColl)
+	connCounter := freqconn.NewConnCounter(strobeThreshold, strobesNotifier)
+
+	db := DB{
+		manager:         dbManager,
+		outputDB:        outputDB,
+		connColl:        connColl,
+		connCounter:     connCounter,
+		strobesNotifier: strobesNotifier,
+	}
+
+	err := db.ensureConnIndexExists()
+	if err != nil {
+		strobesSess.Close()
+		connSess.Close()
+		return db, err
+	}
+
+	err = db.ensureFreqConnIndexExists()
+	if err != nil {
+		strobesSess.Close()
+		connSess.Close()
+		return db, err
+	}
+
+	started := connColl.StartAutoFlush(asyncErrorChan, onFatalError)
+	if !started {
+		err = errors.Errorf("failed to start auto flusher for collection %s.%s", outputDB.Name, constants.ConnCollection)
+		strobesSess.Close()
+		connSess.Close()
+		return db, err
+	}
+
+	err = dbManager.ensureMetaDBRecordExists(outputDB.Name)
+	if err != nil {
+		strobesSess.Close()
+		connSess.Close()
+		return db, err
+	}
+
+	return db, nil
+}
+
+func (d DB) ensureConnIndexExists() error {
+	tmpConn := parsetypes.Conn{}
+	for _, index := range tmpConn.Indices() {
+		err := d.outputDB.C(constants.ConnCollection).EnsureIndex(mgo.Index{
+			Key: []string{index},
+		})
+
+		if err != nil {
+			return errors.Wrapf(err, "could not create RITA conn index %s", index)
+		}
+	}
+	return nil
+}
+
+func (d DB) ensureFreqConnIndexExists() error {
+	tmpFreq := parsetypes.Freq{}
+	for _, index := range tmpFreq.Indices() {
+		err := d.outputDB.C(constants.StrobesCollection).EnsureIndex(mgo.Index{
+			Key: []string{index},
+		})
+
+		if err != nil {
+			return errors.Wrapf(err, "could not create RITA freqConn index %s", index)
+		}
+	}
+	return nil
+}
+
+//InsertConnRecord writes a connection record to the RITA database.
+//Each connection pair is counted, and if the count exceeds a threshold,
+//the connection info is sent to freqConn, otherwise it is sent to conn
+func (d DB) InsertConnRecord(connRecord *parsetypes.Conn) error {
```

Comment:
This method implements the count check as to whether we should send this record over to the conn collection or drop it.

Reply:
The actual code to insert a freqconn entry is called via Increment and the strobesNotifier